reassembly_queue.go

```diff
@@ -1,116 +1,329 @@
+	// This is an ordered chunk
 
-}
+	//fmt.Printf("SI[%d]: push SSN %d Next SSN=%d\n", r.si, chunk.streamSequenceNumber, r.nextSSN)
+
+	if sna16LT(chunk.streamSequenceNumber, r.nextSSN) {
+		fmt.Printf("SI[%d]: ignore stale SSN %d < %d\n", r.si, chunk.streamSequenceNumber, r.nextSSN)
+		return false
+	}
+
+	// Check if a chunkSet with the SSN already exists
+	for _, set := range r.ordered {
+		if set.ssn == chunk.streamSequenceNumber {
+			cset = set
+			break
+		}
+	}
 
-func (m *payloadDataMessage) clear() {
-	m.length = 0
-	m.fragmentQueue = []*chunkPayloadData{}
+	// If not found, create a new chunkSet
+	if cset == nil {
+		cset = newChunkSet(chunk.streamSequenceNumber, chunk.payloadType)
+		r.ordered = append(r.ordered, cset)
+		if !chunk.unordered {
+			sortChunksBySSN(r.ordered)
+		}
+	}
+
+	return cset.push(chunk)
 }
 
-func (m *payloadDataMessage) assemble() ([]byte, bool) {
-	if m.complete() {
-		b := make([]byte, m.length)
-		i := 0
-		for _, p := range m.fragmentQueue {
-			copy(b[i:], p.userData)
-			i += len(p.userData)
+func (r *reassemblyQueue) findCompleteUnorderedChunkSet() *chunkSet {
+	startIdx := -1
+	nChunks := 0
+	var lastTSN uint32
+	var found bool
+
+	for i, c := range r.unorderedChunks {
+		// seek beigining
+		if c.beginningFragment {
+			startIdx = i
+			nChunks = 1
+			lastTSN = c.tsn
+
+			if c.endingFragment {
+				found = true
+				break
+			}
+			continue
+		}
+
+		if startIdx < 0 {
+			continue
+		}
+
+		// Check if contiguous in TSN
+		if c.tsn != lastTSN+1 {
+			startIdx = -1
+			continue
+		}
+
+		lastTSN = c.tsn
+		nChunks++
+
+		if c.endingFragment {
+			found = true
+			break
 		}
+	}
 
-		return b, true
+	if !found {
+		return nil
 	}
 
-	return nil, false
+	// Extract the range of chunks
+	chunks := r.unorderedChunks[startIdx : startIdx+nChunks]
+	r.unorderedChunks = append(
+		r.unorderedChunks[:startIdx],
+		r.unorderedChunks[startIdx+nChunks:]...)
+
+	chunkSet := newChunkSet(0, chunks[0].payloadType)
+	for _, c := range chunks {
+		chunkSet.push(c)
+	}
+
+	return chunkSet
 }
 
-type reassemblyQueue struct {
-	messageQueue     payloadDataMessageArray
-	unorderedMessage payloadDataMessage
-	expectedSeqNum   uint16
+func (r *reassemblyQueue) isReadable() bool {
+	// Check unordered first
+	if len(r.unordered) > 0 {
+		// The chunk sets in r.unordered should all be complete.
+		return true
+	}
+
+	// Check ordered sets
+	if len(r.ordered) > 0 {
+		cset := r.ordered[0]
+		if cset.isComplete() {
+			if sna16LTE(cset.ssn, r.nextSSN) {
+				return true
+			}
+		}
+	}
+	return false
 }
 
-func (r *reassemblyQueue) push(p *chunkPayloadData) {
-	if p.unordered {
-		r.unorderedMessage.fragmentQueue = append(r.unorderedMessage.fragmentQueue, p)
-		r.unorderedMessage.length += len(p.userData)
-		r.unorderedMessage.payloadType = p.payloadType
-		return
+func (r *reassemblyQueue) read(buf []byte) (int, PayloadProtocolIdentifier, error) {
+	var cset *chunkSet
+	// Check unordered first
```

Comment:
Is there a rule to state which is checked first?

Reply:
I have not come across any spec on that but I just followed how it was implemented (was checking the unordered packet first), and also with my assumption (for 2c), an application would choose to use "unordered" delivery mostly because for smaller latency. I had this impression from the text in RFC 4960 sec 6.6:
```
   IMPLEMENTATION NOTE: When sending an unordered DATA chunk, an
   implementation may choose to place the DATA chunk in an outbound
   packet that is at the head of the outbound transmission queue if
   possible.
```

Reply:
Sounds good! I wasn't sure if it was noted in the RFC specifically (and if so, best to leave a comment reflecting that)

Reply:
Leave a note is a good idea! Will do it.